benches/end_to_end.rs

```diff
@@ -1,385 +1,472 @@
-//         }
-//     });
-//     addr
-// }
+struct Opts {
+    http2: bool,
+    http2_stream_window: Option<u32>,
+    http2_conn_window: Option<u32>,
+    http2_adaptive_window: bool,
+    parallel_cnt: u32,
+    request_method: Method,
+    request_body: Option<&'static [u8]>,
+    request_chunks: usize,
+    response_body: &'static [u8],
+}
+
+fn opts() -> Opts {
+    Opts {
+        http2: false,
+        http2_stream_window: None,
+        http2_conn_window: None,
+        http2_adaptive_window: false,
+        parallel_cnt: 1,
+        request_method: Method::GET,
+        request_body: None,
+        request_chunks: 0,
+        response_body: b"",
+    }
+}
+
+impl Opts {
+    fn http2(mut self) -> Self {
+        self.http2 = true;
+        self
+    }
+
+    fn http2_stream_window(mut self, sz: impl Into<Option<u32>>) -> Self {
+        assert!(!self.http2_adaptive_window);
+        self.http2_stream_window = sz.into();
+        self
+    }
+
+    fn http2_conn_window(mut self, sz: impl Into<Option<u32>>) -> Self {
+        assert!(!self.http2_adaptive_window);
+        self.http2_conn_window = sz.into();
+        self
+    }
+
+    fn http2_adaptive_window(mut self) -> Self {
+        assert!(self.http2_stream_window.is_none());
+        assert!(self.http2_conn_window.is_none());
+        self.http2_adaptive_window = true;
+        self
+    }
+
+    fn method(mut self, m: Method) -> Self {
+        self.request_method = m;
+        self
+    }
+
+    fn request_body(mut self, body: &'static [u8]) -> Self {
+        self.request_body = Some(body);
+        self
+    }
+
+    fn request_chunks(mut self, chunk: &'static [u8], cnt: usize) -> Self {
+        assert!(cnt > 0);
+        self.request_body = Some(chunk);
+        self.request_chunks = cnt;
+        self
+    }
+
+    fn response_body(mut self, body: &'static [u8]) -> Self {
+        self.response_body = body;
+        self
+    }
+
+    fn parallel(mut self, cnt: u32) -> Self {
+        assert!(cnt > 0, "parallel count must be larger than 0");
+        self.parallel_cnt = cnt;
+        self
+    }
+
+    fn bench(self, b: &mut test::Bencher) {
+        let _ = pretty_env_logger::try_init();
+        // Create a runtime of current thread.
+        let rt = Arc::new(
+            tokio::runtime::Builder::new_current_thread()
+                .enable_all()
+                .build()
+                .expect("rt build"),
+        );
+
+        let req_len = self.request_body.map(|b| b.len()).unwrap_or(0) as u64
+            * u64::max(1, self.request_chunks as u64);
+        let bytes_per_iter = (req_len + self.response_body.len() as u64) * self.parallel_cnt as u64;
+        b.bytes = bytes_per_iter;
+
+        let addr = spawn_server(&rt, &self);
+
+        match self.http2 {
+            true => self.bench_http2(b, &rt, &addr),
+            false => self.bench_http1(b, &rt, &addr),
+        };
+    }
+
+    //
+    // Benches http/1 requests
+    //
```

Comment:
In Rust we can write [doc comments](https://doc.rust-lang.org/rust-by-example/meta/doc.html) with `///` :slightly_smiling_face:

Reply:
nice